support-workers/src/main/scala/com/gu/zuora/subscriptionBuilders/SubscriptionBuilder.scala

```diff
@@ -0,0 +1,94 @@
+package com.gu.zuora.subscriptionBuilders
+
+import cats.data.EitherT
+import cats.implicits._
+import com.gu.support.config.{TouchPointEnvironment, ZuoraContributionConfig}
+import com.gu.support.promotions.{PromoError, PromotionService}
+import com.gu.support.redemption.{InvalidCode, InvalidReaderType}
+import com.gu.support.redemptions.RedemptionData
+import com.gu.support.workers._
+import com.gu.support.workers.states.CreateZuoraSubscriptionState
+import com.gu.support.zuora.api.ReaderType.Gift
+import com.gu.support.zuora.api.{ReaderType, SubscriptionData}
+import com.gu.zuora.subscriptionBuilders.ProductSubscriptionBuilders.buildContributionSubscription
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+case class BuildSubscribePromoError(cause: PromoError) extends RuntimeException
+
+case class BuildSubscribeRedemptionError(cause: InvalidCode) extends RuntimeException
+
+class SubscriptionBuilder(
+  digitalSubscriptionPurchaseBuilder: DigitalSubscriptionPurchaseBuilder,
```

Comment:
It's a bit strange that we now have instantiated classes being passed in for the digi sub builders while all the others are static objects. Why did you choose to move to the instantiated class approach for digi sub?

Reply:
discussed on a call, summary is
I didn't want to concern the decision logic with non State specific dependencies e.g. config 
if we pass them in all the way in/out as parameters or as a product type we have to explicit about every single one which makes the signatures more complicated.
if we pass as a partially applied/class, it's only relevant what it can do, not what it contains - the things that went into it are encapsulated/private and can't be used directly.  This is likely to produce a simpler signature.  Although of course functions are not serialisable and there is no way to get back the original components by default.